src/v2020/mod.rs

```diff
@@ -707,6 +711,158 @@ impl<R: Read> Iterator for StreamCDC<R> {
+            eof: false,
+            processed: 0,
+            min_size: min_size as usize,
+            avg_size: avg_size as usize,
+            max_size: max_size as usize,
+            mask_s,
+            mask_l,
+            mask_s_ls: mask_s << 1,
+            mask_l_ls: mask_l << 1,
+        }
+    }
+
+    /// Fill the buffer with data from the source, returning the number of bytes
+    /// read (zero if end of source has been reached).
+    async fn fill_buffer(&mut self) -> Result<usize, Error> {
+        // this code originally copied from asuran crate
+        if self.eof {
+            Ok(0)
+        } else {
+            let mut all_bytes_read = 0;
+            while !self.eof && self.length < self.capacity {
+                let bytes_read = self.source.read(&mut self.buffer[self.length..]).await?;
+                if bytes_read == 0 {
+                    self.eof = true;
+                } else {
+                    self.length += bytes_read;
+                    all_bytes_read += bytes_read;
+                }
+            }
+            Ok(all_bytes_read)
+        }
+    }
+
+    /// Drains a specified number of bytes from the buffer, then resizes the
+    /// buffer back to `capacity` size in preparation for further reads.
+    fn drain_bytes(&mut self, count: usize) -> Result<Vec<u8>, Error> {
+        // this code originally copied from asuran crate
+        if count > self.length {
+            Err(Error::Other(format!(
+                "drain_bytes() called with count larger than length: {} > {}",
+                count, self.length
+            )))
+        } else {
+            let data = self.buffer.drain(..count).collect::<Vec<u8>>();
+            self.length -= count;
+            self.buffer.resize(self.capacity, 0_u8);
+            Ok(data)
+        }
+    }
+
+    /// Find the next chunk in the source. If the end of the source has been
+    /// reached, returns `Error::Empty` as the error.
+    async fn read_chunk(&mut self) -> Result<ChunkData, Error> {
+        self.fill_buffer().await?;
+        if self.length == 0 {
+            Err(Error::Empty)
+        } else {
+            let (hash, count) = cut(
+                &self.buffer[..self.length],
+                self.min_size,
+                self.avg_size,
+                self.max_size,
+                self.mask_s,
+                self.mask_l,
+                self.mask_s_ls,
+                self.mask_l_ls,
+            );
+            if count == 0 {
+                Err(Error::Empty)
+            } else {
+                let offset = self.processed;
+                self.processed += count as u64;
+                let data = self.drain_bytes(count)?;
+                Ok(ChunkData {
+                    hash,
+                    offset,
+                    length: count,
+                    data,
+                })
+            }
+        }
+    }
+
+    pub fn as_stream(&mut self) -> impl Stream<Item = Result<ChunkData, Error>> + '_ {
```

Comment:
Tried to implement `Stream` trait on `AsyncStreamCDC` but that was tricky with methods being async already.